File: Recursion/queens.c
/******************************************************************************
 *
 *  Print all solutions to the N queens problem
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>

void queensRecursion(int n, int *array, int i) {
    if (i == n) {
        for(int j = 0; j < n; ++j)
            printf("+-");
        printf("+\n");
        for(int k = 0; k < n; ++k) {
            for(int j = 0; j < n; ++j) 
                printf("|%c", array[j] == k ? '*' : ' ');
            printf("|\n");
            for(int j = 0; j < n; ++j)
                printf("+-");
            printf("+\n");
        }
        printf("\n");
         
        return;
    }

    for(int j = 0; j < n; ++j) {
        array[i] = j;
        int k;
        for(k = 0; k < i; ++k) {
            if (array[k] == array[i])
                break;
            if (array[k] - array[i] == k - i)
                break;
            if (array[k] - array[i] == i - k)
                break;
        }
        if (k != i)
            continue;
        queensRecursion(n, array, i + 1);
    } 
}

void queens(int n) {
    int *array = malloc(n * sizeof(int));
    queensRecursion(n, array, 0);
    free(array);
}

int main(int argc, char **argv) {
    queens(atoi(argv[1]));
    return 0;
} 

  

File: Recursion/Queens.java
/******************************************************************************
 *
 *  Print all solutions to the N queens problem
 *
 *****************************************************************************/
public class Queens {
    static void queensRecursion(int[] array, int i) {
        if (i == array.length) {
            for (int e : array)
                System.out.print("+-");
            System.out.println("+");
            for(int k = 0; k < array.length; ++k) {
                for (int e : array)
                    System.out.print("|" + (e == k ? '*' : ' '));
                System.out.println("|");
                for (int e : array)
                    System.out.print("+-");
                System.out.println("+");
            }
            System.out.println();

            return;
        }

        for(int j = 0; j < array.length; ++j) {
            array[i] = j;
            if (isValid(array, i))
                queensRecursion(array, i + 1);
        }
    }

    static boolean isValid(int[] array, int i) {
        int k;
        for(k = 0; k < i; ++k) {
            if (array[k] == array[i])
                return false;
            if (array[k] - array[i] == k - i)
                return false;
            if (array[k] - array[i] == i - k)
                return false;
        }
        return true;
    }

    static void queens(int n) {
        int[] array = new int[n];
        queensRecursion(array, 0);
    }

    public static void main(String[] args) {
        queens(5);
    }
}
